program/src/processor.rs

```diff
@@ -820,6 +820,54 @@ fn process_switch_vote(
     Ok(())
 }
 
+/// Processes a
+/// [FinishVoting](enum.PaladinGovernanceInstruction.html)
+/// instruction.
+fn process_finish_voting(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let proposal_info = next_account_info(accounts_iter)?;
+
+    check_proposal_exists(program_id, proposal_info)?;
+
+    let mut proposal_data = proposal_info.try_borrow_mut_data()?;
+    let proposal_state = bytemuck::try_from_bytes_mut::<Proposal>(&mut proposal_data)
+        .map_err(|_| ProgramError::InvalidAccountData)?;
+
+    // Ensure the proposal is in the voting stage.
+    if proposal_state.status != ProposalStatus::Voting {
+        return Err(PaladinGovernanceError::ProposalNotInVotingStage.into());
+    }
+
+    let clock = <Clock as Sysvar>::get()?;
+
+    match proposal_state.cooldown_timestamp {
+        Some(_) => {
+            // If the proposal is in a cooldown period, check if it has ended.
+            if proposal_state.cooldown_has_ended(&clock) {
+                // If the cooldown period has ended, the proposal is accepted.
+                proposal_state.status = ProposalStatus::Accepted;
+                Ok(())
+            } else {
+                // If not, the proposal remains in the voting stage.
+                Err(PaladinGovernanceError::ProposalNotAccepted.into())
```

Comment:
nit: how about returning `ProposalVotingPeriodStillActive` or creating a new error like `ProposalCooldownPeriodActive`?

Reply:
Yeah, under the logic of "cooldowns always end in acceptance unless rejection threshold is met", this error made sense. However, considering the other comment, I've changed to `ProposalVotingPeriodStillActive`.